### ISSUE TITLE
multiple button press to get one output

### ISSUE THREAD

**Issue by NerdsCorp** [atctwo/bbrx#2]
I might be missing it or this might be a future thing, but is it possible to Hold one button down and then click another for the outcome to trigger? 

I tried this below but I don't think its built in to the rest of the code. But it would be cool if this was added!

- action: BB_ACTION_SPEED_DOWN
  event: BB_EVENT_DPAD_DOWN
  event: BB_EVENT_BTN_START
  min: 0
  max: 1

**Comment by NerdsCorp** [comment]
this would make it so you could free up main buttons and have "config/tuning" in a different layer thus creating more advanced Projects using this.

**Comment by atctwo** [comment]
I think i might have an idea of how to do this.  What if each binding had an optional key called `conditionals` or something, which can either be an event or a list of events.  The binding's action is only executed if all of the conditional events resolve as true.  I think each event would be resolved by the same logic as boolean-input actions (like BB_ACTION_SPEED_DOWN), where it is `true` if the event value is greater than the halfway point between min and max.  Because the main event and conditional events might have different ranges, I'll also add a `conditional_min` (default=0) and `conditional_max` (default=1).

For example, if you wanted to have speed control stuff gated behind the start button (like in your post) you could do something like this:
```yaml
- action: BB_ACTION_SPEED_DOWN
  event: BB_EVENT_DPAD_DOWN
  conditionals: BB_EVENT_BTN_START
  min: 0
  max: 1
```

Another example, if you instead wanted to check that _both_ analog sticks are pushed upwards (ie: Y axis > 0), you could do
```yaml
- action: BB_ACTION_SPEED_DOWN
  event: BB_EVENT_DPAD_DOWN
  conditionals: [BB_EVENT_ANALOG_LY, BB_EVENT_ANALOG_RY]
  min: 0
  max: 1
  conditional_min: 0
  conditional_max: 511
```

At least I think this would work?  What do you think?

**Comment by NerdsCorp** [comment]
I think that would make the code Clean and it would make it work. Using the conditionals would also allow for the feedback system that is in the #todo.md.

**Comment by atctwo** [comment]
hi, i've added this feature, and a little bit of documentation.  try it out and let me know if it works!

**Comment by NerdsCorp** [comment]
So i did some testing, I tried it a few different ways i thought it would work but nothing. I only changed one, the very last on my config.yml. The way i understand it this config should have two states on and off, its assigned to x but you have to hold the brake, am i right? Im just using an led on that output pin to test.

Also my questions are from reading the Documents are... Do you have to have conditional min and max for them to work or is it alway 1 or 0, its not very clear... and what are possible limitations with conditionals.


Side note, maybe in the future could we clean up the debug codes, its a little
 hard to follow.

Heres the output of my serial.

23:22:44.867 -> Bluepad32 (C) 2016-2024 Ricardo Quesada and contributors.
23:22:44.867 -> Version: v4.1.0
23:22:44.867 -> BTstack: Copyright (C) 2017 BlueKitchen GmbH.
23:22:45.495 -> BTstack up and running at B0:B2:1C:F8:73:AE
23:22:47.016 -> [32m[i] [0m[main] 
23:22:47.016 -> === bbrx! ===
23:22:47.016 -> [32m[i] [0m[main] version:     v1.1
23:22:47.016 -> [32m[i] [0m[main] build time:  Sep 24 2024 23:04:39
23:22:47.016 -> [32m[i] [0m[main] 
23:22:47.016 -> [32m[i] [0m[status] Setting up Status LED
23:22:47.016 -> [36m[d] [0m[status] [2m1 leds on pin 5, init brightness = 35[0m
23:22:47.049 -> [36m[d] [0m[status] [2mchanged state to 1[0m
23:22:47.049 -> [36m[d] [0m[status] [2msetting led to 0 128 0, pulse period = 1000, pulse delay = 4, divider = 7, brightness = 35[0m
23:22:47.049 -> [32m[i] [0m[config] Loading bbrx config file from LittleFS...
23:22:47.115 -> [36m[d] [0m[config] [2mloading deadzones...[0m
23:22:47.115 -> [36m[d] [0m[config] [2m- ly = 40[0m
23:22:47.115 -> [36m[d] [0m[config] [2m- lx = 40[0m
23:22:47.115 -> [36m[d] [0m[config] [2m- ry = 40[0m
23:22:47.115 -> [36m[d] [0m[config] [2m- rx = 40[0m
23:22:47.115 -> [36m[d] [0m[config] [2m- brake = 64[0m
23:22:47.115 -> [36m[d] [0m[config] [2m- throttle = 64[0m
23:22:47.148 -> [36m[d] [0m[config] [2m[0m
23:22:47.148 -> [36m[d] [0m[config] [2mloading beefzones...[0m
23:22:47.148 -> [36m[d] [0m[config] [2m- ly = 500[0m
23:22:47.148 -> [36m[d] [0m[config] [2m- lx = 500[0m
23:22:47.148 -> [36m[d] [0m[config] [2m- ry = 500[0m
23:22:47.148 -> [36m[d] [0m[config] [2m- rx = 500[0m
23:22:47.148 -> [36m[d] [0m[config] [2m- brake = 1000[0m
23:22:47.148 -> [36m[d] [0m[config] [2m- throttle = 1000[0m
23:22:47.148 -> [36m[d] [0m[config] [2m[0m
23:22:47.148 -> [36m[d] [0m[config] [2mparsing binding 0[0m
23:22:47.148 -> [36m[d] [0m[config] [2m- action string BB_ACTION_SERVO[0m
23:22:47.181 -> [36m[d] [0m[config] [2m- action int 2[0m
23:22:47.181 -> [36m[d] [0m[config] [2m- action enum 2[0m
23:22:47.181 -> [36m[d] [0m[config] [2m- event string BB_EVENT_ANALOG_LY[0m
23:22:47.181 -> [36m[d] [0m[config] [2m- event int 1[0m
23:22:47.181 -> [36m[d] [0m[config] [2m- event enum 1[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- min int -512[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- max int 511[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- pin int 36[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.214 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.214 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 0[0m
23:22:47.247 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 0[0m
23:22:47.247 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 0[0m
23:22:47.247 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.281 -> [36m[d] [0m[config] [2m[0m
23:22:47.281 -> [36m[d] [0m[config] [2mparsing binding 1[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- action string BB_ACTION_SERVO[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- action int 2[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- action enum 2[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- event string BB_EVENT_ANALOG_LX[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- event int 0[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- event enum 0[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- min int 511[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- max int -512[0m
23:22:47.281 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.314 -> [36m[d] [0m[config] [2m- pin int 37[0m
23:22:47.314 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.314 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.314 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.347 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.347 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 1[0m
23:22:47.347 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 1[0m
23:22:47.347 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 1[0m
23:22:47.347 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.347 -> [36m[d] [0m[config] [2m[0m
23:22:47.347 -> [36m[d] [0m[config] [2mparsing binding 2[0m
23:22:47.380 -> [36m[d] [0m[config] [2m- action string BB_ACTION_SERVO[0m
23:22:47.380 -> [36m[d] [0m[config] [2m- action int 2[0m
23:22:47.380 -> [36m[d] [0m[config] [2m- action enum 2[0m
23:22:47.380 -> [36m[d] [0m[config] [2m- event string BB_EVENT_ANALOG_RY[0m
23:22:47.380 -> [36m[d] [0m[config] [2m- event int 3[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- event enum 3[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- min int -512[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- max int 511[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- pin int 38[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.413 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.446 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 2[0m
23:22:47.446 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 2[0m
23:22:47.446 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 2[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.479 -> [36m[d] [0m[config] [2m[0m
23:22:47.479 -> [36m[d] [0m[config] [2mparsing binding 3[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- action string BB_ACTION_SERVO[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- action int 2[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- action enum 2[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- event string BB_EVENT_ANALOG_RX[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- event int 2[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- event enum 2[0m
23:22:47.479 -> [36m[d] [0m[config] [2m- min int 511[0m
23:22:47.512 -> [36m[d] [0m[config] [2m- max int -512[0m
23:22:47.512 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.512 -> [36m[d] [0m[config] [2m- pin int 39[0m
23:22:47.512 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.512 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.545 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.545 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.545 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 3[0m
23:22:47.545 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 3[0m
23:22:47.545 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 3[0m
23:22:47.545 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.545 -> [36m[d] [0m[config] [2m[0m
23:22:47.578 -> [36m[d] [0m[config] [2mparsing binding 4[0m
23:22:47.578 -> [36m[d] [0m[config] [2m- action string BB_ACTION_SERVO[0m
23:22:47.578 -> [36m[d] [0m[config] [2m- action int 2[0m
23:22:47.578 -> [36m[d] [0m[config] [2m- action enum 2[0m
23:22:47.578 -> [36m[d] [0m[config] [2m- event string BB_EVENT_ANALOG_THROTTLE[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- event int 5[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- event enum 5[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- min int 1023[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- max int -1024[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- pin int 34[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.611 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.644 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.644 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 4[0m
23:22:47.644 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 4[0m
23:22:47.677 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 4[0m
23:22:47.677 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.677 -> [36m[d] [0m[config] [2m[0m
23:22:47.677 -> [36m[d] [0m[config] [2mparsing binding 5[0m
23:22:47.677 -> [36m[d] [0m[config] [2m- action string BB_ACTION_SERVO[0m
23:22:47.677 -> [36m[d] [0m[config] [2m- action int 2[0m
23:22:47.677 -> [36m[d] [0m[config] [2m- action enum 2[0m
23:22:47.677 -> [36m[d] [0m[config] [2m- event string BB_EVENT_ANALOG_BRAKE[0m
23:22:47.677 -> [36m[d] [0m[config] [2m- event int 4[0m
23:22:47.677 -> [36m[d] [0m[config] [2m- event enum 4[0m
23:22:47.710 -> [36m[d] [0m[config] [2m- min int 1023[0m
23:22:47.710 -> [36m[d] [0m[config] [2m- max int -1024[0m
23:22:47.710 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.710 -> [36m[d] [0m[config] [2m- pin int 35[0m
23:22:47.710 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.743 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.743 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.743 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.743 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 5[0m
23:22:47.743 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 5[0m
23:22:47.743 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 5[0m
23:22:47.776 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.776 -> [36m[d] [0m[config] [2m[0m
23:22:47.776 -> [36m[d] [0m[config] [2mparsing binding 6[0m
23:22:47.776 -> [36m[d] [0m[config] [2m- action string BB_ACTION_GPIO[0m
23:22:47.776 -> [36m[d] [0m[config] [2m- action int 7[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- action enum 7[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- event string BB_EVENT_BTN_B[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- event int 17[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- event enum 17[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- min int 0[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- max int 1[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- pin int 2[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.809 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.842 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.842 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.842 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 6[0m
23:22:47.876 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 6[0m
23:22:47.876 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 6[0m
23:22:47.876 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.876 -> [36m[d] [0m[config] [2m[0m
23:22:47.876 -> [36m[d] [0m[config] [2mparsing binding 7[0m
23:22:47.876 -> [36m[d] [0m[config] [2m- action string BB_ACTION_GPIO[0m
23:22:47.876 -> [36m[d] [0m[config] [2m- action int 7[0m
23:22:47.876 -> [36m[d] [0m[config] [2m- action enum 7[0m
23:22:47.876 -> [36m[d] [0m[config] [2m- event string BB_EVENT_BTN_Y[0m
23:22:47.909 -> [36m[d] [0m[config] [2m- event int 19[0m
23:22:47.909 -> [36m[d] [0m[config] [2m- event enum 19[0m
23:22:47.909 -> [36m[d] [0m[config] [2m- min int 0[0m
23:22:47.909 -> [36m[d] [0m[config] [2m- max int 1[0m
23:22:47.909 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:47.909 -> [36m[d] [0m[config] [2m- pin int 4[0m
23:22:47.942 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:47.942 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:47.942 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:47.942 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:47.942 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 7[0m
23:22:47.942 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 7[0m
23:22:47.942 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 7[0m
23:22:47.975 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:47.975 -> [36m[d] [0m[config] [2m[0m
23:22:47.975 -> [36m[d] [0m[config] [2mparsing binding 8[0m
23:22:47.975 -> [36m[d] [0m[config] [2m- action string BB_ACTION_GPIO[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- action int 7[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- action enum 7[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- event string BB_EVENT_BTN_A[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- event int 16[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- event enum 16[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- min int 0[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- max int 1[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- pin int 17[0m
23:22:48.008 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:48.041 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:48.041 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<string> key[0m
23:22:48.041 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:48.041 -> [36m[d] [0m[config] [2mmissing or invalid conditional_min key in binding 8[0m
23:22:48.074 -> [36m[d] [0m[config] [2mmissing or invalid conditional_max key in binding 8[0m
23:22:48.074 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 8[0m
23:22:48.074 -> [36m[d] [0m[config] [2m- adding binding[0m
23:22:48.074 -> [36m[d] [0m[config] [2m[0m
23:22:48.074 -> [36m[d] [0m[config] [2mparsing binding 9[0m
23:22:48.074 -> [36m[d] [0m[config] [2m- action string BB_ACTION_GPIO[0m
23:22:48.074 -> [36m[d] [0m[config] [2m- action int 7[0m
23:22:48.074 -> [36m[d] [0m[config] [2m- action enum 7[0m
23:22:48.074 -> [36m[d] [0m[config] [2m- event string BB_EVENT_BTN_X[0m
23:22:48.107 -> [36m[d] [0m[config] [2m- event int 18[0m
23:22:48.107 -> [36m[d] [0m[config] [2m- event enum 18[0m
23:22:48.107 -> [36m[d] [0m[config] [2m- min int 0[0m
23:22:48.107 -> [36m[d] [0m[config] [2m- max int 1[0m
23:22:48.107 -> [36m[d] [0m[config] [2m- missing or invalid default_value key[0m
23:22:48.140 -> [36m[d] [0m[config] [2m- pin int 16[0m
23:22:48.140 -> [36m[d] [0m[config] [2m- missing or invalid exec_without_controller key[0m
23:22:48.140 -> [36m[d] [0m[config] [2m- missing or invalid ignore_claims key[0m
23:22:48.140 -> [36m[d] [0m[config] [2m- conditional event string BB_ACTION_BRAKE[0m
23:22:48.140 -> [36m[d] [0m[config] [2m- conditional event int -1[0m
23:22:48.140 -> [33m[w] [0m[config] invalid event for conditionals key in binding 9
23:22:48.140 -> [36m[d] [0m[config] [2m- missing or invalid conditionals<sequence> key[0m
23:22:48.140 -> [36m[d] [0m[config] [2m- conditional_min int 0[0m
23:22:48.173 -> [36m[d] [0m[config] [2m- conditional_max int 1[0m
23:22:48.173 -> [36m[d] [0m[config] [2mmissing or invalid conditional_noexec key in binding 9[0m
23:22:48.173 -> [36m[d] [0m[config] [2m[0m
23:22:48.173 -> [32m[i] [0m[config] Loaded config from LittleFS!
23:22:48.206 -> [32m[i] [0m[controller] Setting up Bluepad32
23:22:48.206 -> [32m[i] [0m[controller] BP32 version: Bluepad32 for Arduino v4.1.0
23:22:48.206 -> [32m[i] [0m[controller] Listening for controllers...
23:22:48.206 -> [32m[i] [0m[events] Initialised binding 0: action=2, event=1, min=-512, max=511
23:22:48.206 -> [32m[i] [0m[events] Initialised binding 1: action=2, event=0, min=511, max=-512
23:22:48.206 -> [32m[i] [0m[events] Initialised binding 2: action=2, event=3, min=-512, max=511
23:22:48.206 -> [32m[i] [0m[events] Initialised binding 3: action=2, event=2, min=511, max=-512
23:22:48.239 -> [32m[i] [0m[events] Initialised binding 4: action=2, event=5, min=1023, max=-1024
23:22:48.239 -> [32m[i] [0m[events] Initialised binding 5: action=2, event=4, min=1023, max=-1024
23:22:48.239 -> [32m[i] [0m[events] Initialised binding 6: action=7, event=17, min=0, max=1
23:22:48.239 -> [32m[i] [0m[events] Initialised binding 7: action=7, event=19, min=0, max=1
23:22:48.272 -> [32m[i] [0m[events] Initialised binding 8: action=7, event=16, min=0, max=1
23:22:48.272 -> [36m[d] [0m[status] [2mchanged state to 2[0m
23:22:48.272 -> [36m[d] [0m[status] [2msetting led to 0 0 255, pulse period = 1000, pulse delay = 4, divider = 7, brightness = 35[0m
23:22:48.272 -> [32m[i] [0m[main] Setup complete!
23:23:05.761 -> [32m[i] [0m[controller] Connected to a controller!
23:23:05.761 -> [32m[i] [0m[controller]   - model:   DualShock 4
23:23:05.761 -> [32m[i] [0m[controller] DS4: Failed to create virtual device
23:23:05.761 ->   - battery: 0%
23:23:05.761 -> [36m[d] [0m[controller] [2mBTAddr: 90:89:5f:aa:a5:b3, VID/PID: 054c:09cc, flags: 0x05[0m
23:23:05.761 -> [36m[d] [0m[status] [2mchanged state to 3[0m
23:23:05.761 -> [36m[d] [0m[status] [2msetting led to 0 255 255, pulse period = -1, pulse delay = -1, divider = 7, brightness = 35[0m
23:23:10.521 -> [36m[d] [0m[events] [2mAction 2 on pin 34 claimed by binding 4[0m
23:23:11.480 -> [36m[d] [0m[events] [2mAction 7 on pin 17 claimed by binding 8[0m
23:23:11.645 -> [36m[d] [0m[events] [2mAction 7 on pin 17 unclaimed by binding 8[0m
23:23:12.108 -> [36m[d] [0m[events] [2mAction 2 on pin 34 unclaimed by binding 4[0m
23:23:12.934 -> [36m[d] [0m[events] [2mAction 7 on pin 17 claimed by binding 8[0m
23:23:13.166 -> [36m[d] [0m[events] [2mAction 7 on pin 17 unclaimed by binding 8[0m
23:23:13.463 -> [36m[d] [0m[events] [2mAction 7 on pin 17 claimed by binding 8[0m
23:23:13.728 -> [36m[d] [0m[events] [2mAction 7 on pin 17 unclaimed by binding 8[0m
23:23:14.323 -> [36m[d] [0m[events] [2mAction 7 on pin 17 claimed by binding 8[0m
23:23:14.422 -> [36m[d] [0m[events] [2mAction 7 on pin 17 unclaimed by binding 8[0m
23:23:15.248 -> [36m[d] [0m[events] [2mAction 7 on pin 17 claimed by binding 8[0m
23:23:15.381 -> [36m[d] [0m[events] [2mAction 7 on pin 17 unclaimed by binding 8[0m
23:23:17.794 -> [36m[d] [0m[events] [2mAction 2 on pin 34 claimed by binding 4[0m
23:23:18.422 -> [36m[d] [0m[events] [2mAction 2 on pin 34 unclaimed by binding 4[0m
23:23:18.852 -> [36m[d] [0m[events] [2mAction 2 on pin 34 claimed by binding 4[0m
23:23:19.282 -> [36m[d] [0m[events] [2mAction 2 on pin 34 unclaimed by binding 4[0m
23:23:19.678 -> [36m[d] [0m[events] [2mAction 2 on pin 34 claimed by binding 4[0m
23:23:20.207 -> [36m[d] [0m[events] [2mAction 7 on pin 17 claimed by binding 8[0m
23:23:20.373 -> [36m[d] [0m[events] [2mAction 7 on pin 17 unclaimed by binding 8[0m
23:23:20.637 -> [36m[d] [0m[events] [2mAction 7 on pin 17 claimed by binding 8[0m
23:23:20.802 -> [36m[d] [0m[events] [2mAction 7 on pin 17 unclaimed by binding 8[0m
23:23:20.901 -> [36m[d] [0m[events] [2mAction 2 on pin 34 unclaimed by binding 4[0m


Heres My Config.yml

deadzones:
  ly: 40
  lx: 40
  ry: 40
  rx: 40
  brake: 64
  throttle: 64

beefzones:
  ly: 500
  lx: 500
  ry: 500
  rx: 500
  brake: 1000
  throttle: 1000

bindings:

- action: BB_ACTION_SERVO
  event: BB_EVENT_ANALOG_LY
  min: -512
  max: 511
  pin: 36


- action: BB_ACTION_SERVO
  event: BB_EVENT_ANALOG_LX
  min: 511
  max: -512
  pin: 37


- action: BB_ACTION_SERVO
  event: BB_EVENT_ANALOG_RY
  min: -512
  max: 511
  pin: 38


- action: BB_ACTION_SERVO
  event: BB_EVENT_ANALOG_RX
  min: 511
  max: -512
  pin: 39


- action: BB_ACTION_SERVO
  event: BB_EVENT_ANALOG_THROTTLE
  min: 1023
  max: -1024
  pin: 34


- action: BB_ACTION_SERVO
  event: BB_EVENT_ANALOG_BRAKE
  min: 1023
  max: -1024
  pin: 35


- action: BB_ACTION_GPIO
  event: BB_EVENT_BTN_B
  min: 0
  max: 1
  pin: 2


- action: BB_ACTION_GPIO
  event: BB_EVENT_BTN_Y
  min: 0
  max: 1
  pin: 4

- action: BB_ACTION_GPIO
  event: BB_EVENT_BTN_A
  min: 0
  max: 1
  pin: 17

- action: BB_ACTION_GPIO
  event: BB_EVENT_BTN_X
  conditionals: BB_ACTION_BRAKE
  min: 0
  max: 1
  pin: 16

**Comment by NerdsCorp** [comment]
Never mind its not an issue, when i moved it from the brake to a different button it worked great. ( may put this in the Notes)= conditionals cannot be a analog option on the controller.

also I have verified "sticky Keys" does not work, and we should add a easy Is sticky in the config.yml like below. I think if its needed it can be turned on otherwise it off to eliminate Nooby Confusion.

action: BB_ACTION_GPIO
event: BB_EVENT_BTN_A
min: 0
max: 1
is_sticky: true or false
pin: 17

**Comment by atctwo** [comment]
thanks for trying it out!
- `conditional_min` is 0 by default and `conditional_max` is 1 by default.  they're designed for digital buttons by default.  good point about it not being documented, i'll add a note in the docs.
- you _can_ use an analog input as a conditional, but you have to set `conditional_min` and `conditional_max` to match the range of the input since 0 and 1 usually aren't appropriate for analog inputs.  again i get that this isn't clear in the docs so i'll add this too
- one limit with conditionals is that you can't mix and match digital and analog _conditional_ inputs, since `conditional_{min, max}` applies to all conditionals.  note you can have a digital `event` and analog `conditionals` or vice versa, since `event` and `conditionals` have separate min and max settings
- since conditionals are basically boolean at the end of the day, if you want to use the brake trigger as a conditional you can just use `BB_EVENT_BTN_L2`.  with `BB_EVENT_ANALOG_BRAKE` you have to set `conditonal_{min, max}` since it's an analog input, but with `BB_EVENT_BTN_L2` you can use the default range since it's digital
- i'm not really sure what you mean by "sticky keys", are you talking about the "latch" feature?  that was actually completely accidental, i didn't actually mean for that feature to exist but i kept it in incase someone finds it useful.  to get it to work you have to let go of the conditional inputs before you let go of the main event input.  it's a bit hacky, but i'm thinking of adding a "toggle switch" feature one day which toggles an action on or off based on input.  but that's a future idea
- regarding the funky debug logs, the `�[36m` stuff is ANSI colour codes.  so on terminals which support it parts of the output will render in colour.  i do most of my development of bbrx in visual studio code, which looks like this:
  
  ![screenshot showing the output of bbrx, where different log levels are rendered using different colours](https://github.com/user-attachments/assets/be12a29f-a187-4da6-893c-518792785b61)
  
  I tried it on arduino ide v2 and it looks like it doesn't support colour codes :(  
  
  ![image](https://github.com/user-attachments/assets/13c40183-90a4-4065-8a36-1eaea23a71db)
  
  there's not really an ideal way to copy and paste the text without ansi codes from arduino ide, but you can disable ansi codes entirely by commenting out [`ANSI_ENABLE`](https://github.com/atctwo/bbrx/blob/main/bbrx/log.h#L15) in log.h

I hope this clears it up!

**Comment by NerdsCorp** [comment]
yes this explanation helps a lot! thanks!
